File: Sources/COCHelperApp/SnapshotImportTransaction.swift
import Foundation
import COCHelperCore

enum CurrentVillageDataValidationError: Error, LocalizedError {
    case invalid(String)

    var errorDescription: String? {
        switch self {
        case .invalid(let message):
            message
        }
    }
}

enum CurrentVillageDataValidator {
    /// Validates the persisted villages blob before a transaction can restore
    /// or replay it.  A legal empty array is distinct from a missing store and
    /// must not be turned into a decode failure.
    static func validate(_ data: Data?, label: String) throws {
        do {
            try VillageStoreCodec.validate(data, label: label)
            guard let data else { return }
            let villages = try JSONDecoder().decode([VillageProfile].self, from: data)
            guard Set(villages.map(\.id)).count == villages.count else {
                throw CurrentVillageDataValidationError.invalid(
                    "\(label) 包含重复的村庄 ID。"
                )
            }
        } catch let error as CurrentVillageDataValidationError {
            throw error
        } catch {
            if let localized = error as? LocalizedError,
               let description = localized.errorDescription {
                throw CurrentVillageDataValidationError.invalid(description)
            }
            throw CurrentVillageDataValidationError.invalid(error.localizedDescription)
        }
    }
}

enum SnapshotHistoryDataValidator {
    static func validate(_ data: Data?, label: String) throws {
        guard let data else { return }
        do {
            let envelope = try JSONDecoder().decode(SnapshotHistoryEnvelope.self, from: data)
            _ = try envelope.validated()
        } catch {
            throw SnapshotImportTransactionError.journalCorrupt(
                "\(label) 无效：\(error.localizedDescription)"
            )
        }
    }
}

enum ManualTrackerDataValidator {
    static func validate(_ data: Data?, label: String) throws {
        guard let data else { return }
        do {
            let envelope = try JSONDecoder().decode(ManualTrackerEnvelope.self, from: data)
            _ = try envelope.validated()
        } catch {
            throw SnapshotImportTransactionError.journalCorrupt(
                "\(label) 无效：\(error.localizedDescription)"
            )
        }
    }
}

/// The current village blob is kept behind a tiny injectable adapter so the
/// import transaction can test partial writes without coupling Core to
/// UserDefaults.
protocol CurrentVillagePersistence {
    func readData() -> Data?
    func writeData(_ data: Data) throws
    func restoreData(_ data: Data?) throws
}

struct UserDefaultsCurrentVillagePersistence: CurrentVillagePersistence {
    let defaults: UserDefaults
    let key: String

    func readData() -> Data? {
        defaults.data(forKey: key)
    }

    func writeData(_ data: Data) throws {
        defaults.set(data, forKey: key)
        guard defaults.data(forKey: key) == data else {
            throw VillageStoreError.writeFailed("写入后读回的村庄数据与候选数据不一致。")
        }
    }

    func restoreData(_ data: Data?) throws {
        if let data {
            defaults.set(data, forKey: key)
            guard defaults.data(forKey: key) == data else {
                throw VillageStoreError.writeFailed("恢复后读回的村庄数据与原始数据不一致。")
            }
        } else {
            defaults.removeObject(forKey: key)
            guard defaults.data(forKey: key) == nil else {
                throw VillageStoreError.writeFailed("删除村庄数据后仍能读到旧数据。")
            }
        }
    }
}

private enum SnapshotImportJournalPhase: String, Codable {
    case prepared
    case committed
}

private struct SnapshotImportJournal: Codable {
    let phase: SnapshotImportJournalPhase
    let previousCurrentData: Data?
    let newCurrentData: Data
    let previousHistoryData: Data?
    let newHistoryData: Data
    let previousManualData: Data?
    /// Explicitly distinguishes "no manual transaction" from a journal that
    /// claims to include one but lost its payload.
    let manualIncluded: Bool
    let newManualData: Data?

    init(
        phase: SnapshotImportJournalPhase,
        previousCurrentData: Data?,
        newCurrentData: Data,
        previousHistoryData: Data?,
        newHistoryData: Data,
        previousManualData: Data? = nil,
        manualIncluded: Bool = false,
        newManualData: Data? = nil
    ) {
        self.phase = phase
        self.previousCurrentData = previousCurrentData
        self.newCurrentData = newCurrentData
        self.previousHistoryData = previousHistoryData
        self.newHistoryData = newHistoryData
        self.previousManualData = previousManualData
        self.manualIncluded = manualIncluded
        self.newManualData = newManualData
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let newManualData = try container.decodeIfPresent(Data.self, forKey: .newManualData)
        self.init(
            phase: try container.decode(SnapshotImportJournalPhase.self, forKey: .phase),
            previousCurrentData: try container.decodeIfPresent(Data.self, forKey: .previousCurrentData),
            newCurrentData: try container.decode(Data.self, forKey: .newCurrentData),
            previousHistoryData: try container.decodeIfPresent(Data.self, forKey: .previousHistoryData),
            newHistoryData: try container.decode(Data.self, forKey: .newHistoryData),
            previousManualData: try container.decodeIfPresent(Data.self, forKey: .previousManualData),
            manualIncluded: try container.decodeIfPresent(Bool.self, forKey: .manualIncluded)
                ?? (newManualData != nil),
            newManualData: newManualData
        )
    }

    private enum CodingKeys: String, CodingKey {
        case phase
        case previousCurrentData
        case newCurrentData
        case previousHistoryData
        case newHistoryData
        case previousManualData
        case manualIncluded
        case newManualData
    }
}

enum SnapshotImportTransactionError: Error, LocalizedError, Equatable {
    case rollbackFailed(String)
    case journalCorrupt(String)

    var errorDescription: String? {
        switch self {
        case .rollbackFailed(let message):
            "导入事务回滚失败：" + message
        case .journalCorrupt(let message):
            "导入事务记录损坏：" + message
        }
    }
}

/// Coordinates the legacy UserDefaults blob and the standalone history file.
/// The journal makes an interrupted process recover to either the previous
/// state or the complete new state; a normal error is rolled back in-process.
struct SnapshotImportTransactionCoordinator {
    let current: any CurrentVillagePersistence
    let history: any SnapshotHistoryStore
    let journalURL: URL?
    let manual: (any ManualTrackerStore)?

    init(
        current: any CurrentVillagePersistence,
        history: any SnapshotHistoryStore,
        journalURL: URL?,
        manual: (any ManualTrackerStore)? = nil
    ) {
        self.current = current
        self.history = history
        self.journalURL = journalURL
        self.manual = manual
    }

    func recoverIfNeeded() throws {
        guard let journalURL,
              FileManager.default.fileExists(atPath: journalURL.path) else { return }

        let journal: SnapshotImportJournal
        do {
            let data = try Data(contentsOf: journalURL)
            journal = try JSONDecoder().decode(SnapshotImportJournal.self, from: data)
        } catch {
            throw SnapshotImportTransactionError.journalCorrupt(error.localizedDescription)
        }

        guard journal.manualIncluded == (journal.newManualData != nil) else {
            throw SnapshotImportTransactionError.journalCorrupt(
                "事务记录的 manualIncluded 与手动状态 payload 不一致。"
            )
        }
        guard journal.manualIncluded || journal.previousManualData == nil else {
            throw SnapshotImportTransactionError.journalCorrupt(
                "无 manual 事务的记录不应包含 previousManualData。"
            )
        }
        do {
            try CurrentVillageDataValidator.validate(
                journal.previousCurrentData,
                label: "事务记录中的旧当前村庄数据"
            )
            try CurrentVillageDataValidator.validate(
                journal.newCurrentData,
                label: "事务记录中的新当前村庄数据"
            )
            try SnapshotHistoryDataValidator.validate(
                journal.previousHistoryData,
                label: "事务记录中的旧历史"
            )
            try SnapshotHistoryDataValidator.validate(
                journal.newHistoryData,
                label: "事务记录中的新历史"
            )
        } catch {
            if let transactionError = error as? SnapshotImportTransactionError {
                throw transactionError
            }
            throw SnapshotImportTransactionError.journalCorrupt(error.localizedDescription)
        }

        let recoveryManualStore: (any ManualTrackerStore)?
        let recoveryManualData: Data?
        if journal.manualIncluded {
            guard let manual else {
                throw SnapshotImportTransactionError.journalCorrupt(
                    "事务记录包含手动状态，但当前未配置手动存储。"
                )
            }
            guard let newManualData = journal.newManualData else {
                throw SnapshotImportTransactionError.journalCorrupt(
                    "事务记录声明包含手动状态，但缺少 newManualData。"
                )
            }
            try ManualTrackerDataValidator.validate(
                journal.previousManualData,
                label: "事务记录中的旧手动状态"
            )
            try ManualTrackerDataValidator.validate(
                newManualData,
                label: "事务记录中的新手动状态"
            )
            recoveryManualStore = manual
            recoveryManualData = newManualData
        } else {
            recoveryManualStore = nil
            recoveryManualData = nil
        }

        switch journal.phase {
        case .prepared:
            try current.restoreData(journal.previousCurrentData)
            try history.restoreRawData(journal.previousHistoryData)
            if let recoveryManualStore {
                try recoveryManualStore.restoreRawData(journal.previousManualData)
            }
        case .committed:
            try current.writeData(journal.newCurrentData)
            try history.writeRawData(journal.newHistoryData)
            if let recoveryManualStore, let recoveryManualData {
                try recoveryManualStore.writeRawData(recoveryManualData)
            }
        }
        try FileManager.default.removeItem(at: journalURL)
    }

    func commit(
        currentData: Data,
        envelope: SnapshotHistoryEnvelope,
        manualEnvelope: ManualTrackerEnvelope? = nil
    ) throws {
        let newHistoryData = try envelope.encodedData()
        guard let existingHistory = try history.load(), existingHistory.isMigrated else {
            throw SnapshotHistoryStoreError.unavailable("导入前未找到可用的已迁移历史。")
        }
        let previousCurrentData = current.readData()
        let previousHistoryData = try history.readRawData()
        guard manualEnvelope == nil || manual != nil else {
            throw SnapshotImportTransactionError.journalCorrupt(
                "提交手动状态时未配置手动存储。"
            )
        }
        let previousManualData: Data?
        if manualEnvelope != nil {
            previousManualData = try manual?.readRawData()
        } else {
            previousManualData = nil
        }
        let newManualData = try manualEnvelope?.encodedData()
        do {
            try CurrentVillageDataValidator.validate(
                previousCurrentData,
                label: "旧当前村庄数据"
            )
            try CurrentVillageDataValidator.validate(currentData, label: "新当前村庄数据")
            try SnapshotHistoryDataValidator.validate(
                previousHistoryData,
                label: "旧历史"
            )
            try SnapshotHistoryDataValidator.validate(newHistoryData, label: "新历史")
            if manualEnvelope != nil {
                try ManualTrackerDataValidator.validate(
                    previousManualData,
                    label: "旧手动状态"
                )
                try ManualTrackerDataValidator.validate(newManualData, label: "新手动状态")
            }
        } catch {
            if let transactionError = error as? SnapshotImportTransactionError {
                throw transactionError
            }
            throw SnapshotImportTransactionError.journalCorrupt(error.localizedDescription)
        }

        let journal = SnapshotImportJournal(
            phase: .prepared,
            previousCurrentData: previousCurrentData,
            newCurrentData: currentData,
            previousHistoryData: previousHistoryData,
            newHistoryData: newHistoryData,
            previousManualData: previousManualData,
            manualIncluded: manualEnvelope != nil,
            newManualData: newManualData
        )
        try writeJournal(journal)

        do {
            try current.writeData(currentData)
            try history.writeRawData(newHistoryData)
            if let newManualData {
                try manual?.writeRawData(newManualData)
            }
        } catch {
            do {
                try current.restoreData(previousCurrentData)
                try history.restoreRawData(previousHistoryData)
                if newManualData != nil {
                    try manual?.restoreRawData(previousManualData)
                }
                try removeJournalIfPresent()
            } catch {
                throw SnapshotImportTransactionError.rollbackFailed(error.localizedDescription)
            }
            throw error
        }

        guard journalURL != nil else { return }
        do {
            try writeJournal(SnapshotImportJournal(
                phase: .committed,
                previousCurrentData: previousCurrentData,
                newCurrentData: currentData,
                previousHistoryData: previousHistoryData,
                newHistoryData: newHistoryData,
                previousManualData: previousManualData,
                manualIncluded: manualEnvelope != nil,
                newManualData: newManualData
            ))
        } catch {
            do {
                try current.restoreData(previousCurrentData)
                try history.restoreRawData(previousHistoryData)
                if newManualData != nil {
                    try manual?.restoreRawData(previousManualData)
                }
                try removeJournalIfPresent()
            } catch {
                throw SnapshotImportTransactionError.rollbackFailed(error.localizedDescription)
            }
            throw error
        }

        // The committed journal is itself a valid recovery record.  If only
        // cleanup fails, keep the new state and let the next launch replay it
        // idempotently rather than reporting a false import failure.
        try? removeJournalIfPresent()
    }

    private func writeJournal(_ journal: SnapshotImportJournal) throws {
        guard let journalURL else { return }
        do {
            try FileManager.default.createDirectory(
                at: journalURL.deletingLastPathComponent(),
                withIntermediateDirectories: true
            )
            let data = try JSONEncoder().encode(journal)
            try data.write(to: journalURL, options: .atomic)
        } catch {
            throw SnapshotImportTransactionError.journalCorrupt(error.localizedDescription)
        }
    }

    private func removeJournalIfPresent() throws {
        guard let journalURL,
              FileManager.default.fileExists(atPath: journalURL.path) else { return }
        try FileManager.default.removeItem(at: journalURL)
    }
}

File: packages/domain/src/persistence/snapshot-import-transaction.ts
import { existsSync, mkdirSync, readFileSync, rmSync } from 'node:fs';
import { dirname } from 'node:path';

import type { ManualTrackerStore } from '../manual/file-store';
import {
  decodeManualTrackerEnvelopeWire,
  encodeManualTrackerEnvelopeWire,
} from '../manual/tracker-wire';
import type { ManualTrackerEnvelope } from '../manual/tracker-envelope';
import type { SnapshotHistoryStore } from '../snapshot-history/store-port';
import {
  decodeSnapshotHistoryEnvelopeWire,
  encodeSnapshotHistoryEnvelopeWire,
} from '../snapshot-history/envelope-wire';
import { validateSnapshotHistoryEnvelope } from '../snapshot-history/envelope-validate';
import { envelopeIsMigrated, type SnapshotHistoryEnvelope } from '../snapshot-history/store-types';
import { atomicWriteFile } from './atomic-write';
import { base64ToBytes, bytesToBase64 } from './bytes';
import type { WriteFaultInjector } from './fault';
import type { CurrentVillagePersistence } from './village-file-store';
import { validateVillageStoreBytes } from './village-codec';
import type { PerformanceTraceSink } from '../performance';

export type SnapshotImportJournalPhase = 'prepared' | 'committed';

export type SnapshotImportJournalV1 = {
  readonly phase: SnapshotImportJournalPhase;
  readonly previousCurrentData: string | null;
  readonly newCurrentData: string;
  readonly previousHistoryData: string | null;
  readonly newHistoryData: string;
  readonly previousManualData: string | null;
  readonly manualIncluded: boolean;
  readonly newManualData: string | null;
};

export type SnapshotImportTransactionError =
  | { readonly kind: 'rollbackFailed'; readonly message: string }
  | { readonly kind: 'journalCorrupt'; readonly message: string };

export type SnapshotImportTransactionCoordinatorOptions = {
  readonly current: CurrentVillagePersistence;
  readonly history: SnapshotHistoryStore;
  readonly journalURL: string | null;
  readonly manual?: ManualTrackerStore | null;
  readonly fault?: WriteFaultInjector;
  readonly performanceTrace?: PerformanceTraceSink;
};

export class SnapshotImportTransactionCoordinator {
  private readonly current: CurrentVillagePersistence;
  private readonly history: SnapshotHistoryStore;
  private readonly journalURL: string | null;
  private readonly manual: ManualTrackerStore | null;
  private readonly fault: WriteFaultInjector | undefined;
  private readonly performanceTrace: PerformanceTraceSink | undefined;

  constructor(options: SnapshotImportTransactionCoordinatorOptions) {
    this.current = options.current;
    this.history = options.history;
    this.journalURL = options.journalURL;
    this.manual = options.manual ?? null;
    this.fault = options.fault;
    this.performanceTrace = options.performanceTrace;
  }

  recoverIfNeeded(): void {
    if (this.journalURL === null || !existsSync(this.journalURL)) {
      return;
    }

    let journal: SnapshotImportJournalV1;
    try {
      journal = JSON.parse(readFileSync(this.journalURL, 'utf8')) as SnapshotImportJournalV1;
      journal = normalizeImportJournal(journal);
    } catch (error) {
      throw {
        kind: 'journalCorrupt',
        message: error instanceof Error ? error.message : String(error),
      } satisfies SnapshotImportTransactionError;
    }

    if (journal.manualIncluded !== (journal.newManualData !== null)) {
      throw {
        kind: 'journalCorrupt',
        message: '事务记录的 manualIncluded 与手动状态 payload 不一致。',
      } satisfies SnapshotImportTransactionError;
    }
    if (!journal.manualIncluded && journal.previousManualData !== null) {
      throw {
        kind: 'journalCorrupt',
        message: '无 manual 事务的记录不应包含 previousManualData。',
      } satisfies SnapshotImportTransactionError;
    }

    try {
      validateVillageStoreBytes(
        decodeOptionalBase64(journal.previousCurrentData),
        '事务记录中的旧当前村庄数据',
      );
      validateVillageStoreBytes(
        base64ToBytes(journal.newCurrentData),
        '事务记录中的新当前村庄数据',
      );
      validateHistoryBytes(
        decodeOptionalBase64(journal.previousHistoryData),
        '事务记录中的旧历史',
        this.performanceTrace,
        'recovery.validate-previous-history',
      );
      validateHistoryBytes(
        base64ToBytes(journal.newHistoryData),
        '事务记录中的新历史',
        this.performanceTrace,
        'recovery.validate-new-history',
      );
    } catch (error) {
      throw asJournalCorrupt(error);
    }

    let recoveryManualStore: ManualTrackerStore | null = null;
    let recoveryManualData: Uint8Array | null = null;
    if (journal.manualIncluded) {
      if (this.manual === null) {
        throw {
          kind: 'journalCorrupt',
          message: '事务记录包含手动状态，但当前未配置手动存储。',
        } satisfies SnapshotImportTransactionError;
      }
      if (journal.newManualData === null) {
        throw {
          kind: 'journalCorrupt',
          message: '事务记录声明包含手动状态，但缺少 newManualData。',
        } satisfies SnapshotImportTransactionError;
      }
      try {
        validateManualBytes(
          decodeOptionalBase64(journal.previousManualData),
          '事务记录中的旧手动状态',
        );
        validateManualBytes(base64ToBytes(journal.newManualData), '事务记录中的新手动状态');
      } catch (error) {
        throw asJournalCorrupt(error);
      }
      recoveryManualStore = this.manual;
      recoveryManualData = base64ToBytes(journal.newManualData);
    }

    switch (journal.phase) {
      case 'prepared':
        this.current.restoreData(decodeOptionalBase64(journal.previousCurrentData));
        this.history.restoreRawData(decodeOptionalBase64(journal.previousHistoryData));
        if (recoveryManualStore !== null) {
          recoveryManualStore.restoreRawData(decodeOptionalBase64(journal.previousManualData));
        }
        break;
      case 'committed':
        this.current.writeData(base64ToBytes(journal.newCurrentData));
        this.history.writeRawData(base64ToBytes(journal.newHistoryData));
        if (recoveryManualStore !== null && recoveryManualData !== null) {
          recoveryManualStore.writeRawData(recoveryManualData);
        }
        break;
      default:
        throw {
          kind: 'journalCorrupt',
          message: `未知事务阶段：${String((journal as SnapshotImportJournalV1).phase)}`,
        } satisfies SnapshotImportTransactionError;
    }
    rmSync(this.journalURL);
  }

  commit(input: {
    readonly currentData: Uint8Array;
    readonly envelope: SnapshotHistoryEnvelope;
    /** 已由调用方从同一 import 生命周期加载并校验的当前 history。 */
    readonly existingHistory?: SnapshotHistoryEnvelope | null;
    readonly manualEnvelope?: ManualTrackerEnvelope | null;
  }): void {
    let newHistoryData: Uint8Array;
    try {
      newHistoryData = this.measure('history', 'encode', () =>
        new TextEncoder().encode(encodeSnapshotHistoryEnvelopeWire(input.envelope)),
      );
    } catch (error) {
      throw asJournalCorrupt(error);
    }
    const existingHistory = input.existingHistory ?? this.history.load();
    if (existingHistory === null || !envelopeIsMigrated(existingHistory)) {
      throw {
        kind: 'unavailable',
        message: '导入前未找到可用的已迁移历史。',
      };
    }

    const previousCurrentData = this.current.readData();
    const previousHistoryData = this.history.readRawData();
    if (input.manualEnvelope != null && this.manual === null) {
      throw {
        kind: 'journalCorrupt',
        message: '提交手动状态时未配置手动存储。',
      } satisfies SnapshotImportTransactionError;
    }
    const previousManualData = input.manualEnvelope != null ? this.manual!.readRawData() : null;
    const newManualData =
      input.manualEnvelope != null
        ? new TextEncoder().encode(encodeManualTrackerEnvelopeWire(input.manualEnvelope))
        : null;

    try {
      validateVillageStoreBytes(previousCurrentData, '旧当前村庄数据');
      validateVillageStoreBytes(input.currentData, '新当前村庄数据');
      validateHistoryBytes(
        previousHistoryData,
        '旧历史',
        this.performanceTrace,
        'validate-previous-history',
      );
      validateHistoryBytes(
        newHistoryData,
        '新历史',
        this.performanceTrace,
        'validate-wire-history',
      );
      if (input.manualEnvelope != null) {
        validateManualBytes(previousManualData, '旧手动状态');
        validateManualBytes(newManualData, '新手动状态');
      }
    } catch (error) {
      throw asJournalCorrupt(error);
    }

    const prepared: SnapshotImportJournalV1 = {
      phase: 'prepared',
      previousCurrentData: encodeOptionalBase64(previousCurrentData),
      newCurrentData: bytesToBase64(input.currentData),
      previousHistoryData: encodeOptionalBase64(previousHistoryData),
      newHistoryData: bytesToBase64(newHistoryData),
      previousManualData: encodeOptionalBase64(previousManualData),
      manualIncluded: input.manualEnvelope != null,
      newManualData: newManualData === null ? null : bytesToBase64(newManualData),
    };
    this.measure('storage', 'journal-prepared', () => this.writeJournal(prepared));

    try {
      this.measure('storage', 'write', () => {
        this.current.writeData(input.currentData);
        this.history.writeRawData(newHistoryData);
        if (newManualData !== null) {
          this.manual!.writeRawData(newManualData);
        }
      });
    } catch (error) {
      try {
        this.current.restoreData(previousCurrentData);
        this.history.restoreRawData(previousHistoryData);
        if (newManualData !== null) {
          this.manual!.restoreRawData(previousManualData);
        }
        this.removeJournalIfPresent();
      } catch (rollbackError) {
        throw {
          kind: 'rollbackFailed',
          message: rollbackError instanceof Error ? rollbackError.message : String(rollbackError),
        } satisfies SnapshotImportTransactionError;
      }
      throw error;
    }

    if (this.journalURL === null) {
      return;
    }
    try {
      this.measure('storage', 'journal-committed', () =>
        this.writeJournal({
          ...prepared,
          phase: 'committed',
        }),
      );
    } catch (error) {
      try {
        this.current.restoreData(previousCurrentData);
        this.history.restoreRawData(previousHistoryData);
        if (newManualData !== null) {
          this.manual!.restoreRawData(previousManualData);
        }
        this.removeJournalIfPresent();
      } catch (rollbackError) {
        throw {
          kind: 'rollbackFailed',
          message: rollbackError instanceof Error ? rollbackError.message : String(rollbackError),
        } satisfies SnapshotImportTransactionError;
      }
      throw error;
    }

    try {
      this.removeJournalIfPresent();
    } catch {
      // committed journal 本身是合法恢复记录
    }
  }

  private writeJournal(journal: SnapshotImportJournalV1): void {
    if (this.journalURL === null) {
      return;
    }
    try {
      mkdirSync(dirname(this.journalURL), { recursive: true });
      atomicWriteFile(this.journalURL, new TextEncoder().encode(JSON.stringify(journal)), {
        fault: this.fault,
      });
    } catch (error) {
      throw {
        kind: 'journalCorrupt',
        message: error instanceof Error ? error.message : String(error),
      } satisfies SnapshotImportTransactionError;
    }
  }

  private removeJournalIfPresent(): void {
    if (this.journalURL === null || !existsSync(this.journalURL)) {
      return;
    }
    rmSync(this.journalURL);
  }

  private measure<T>(scope: string, phase: string, task: () => T): T {
    return this.performanceTrace === undefined
      ? task()
      : this.performanceTrace.measure(scope, phase, task);
  }
}

function normalizeImportJournal(journal: SnapshotImportJournalV1): SnapshotImportJournalV1 {
  if (typeof journal.manualIncluded === 'boolean') {
    return journal;
  }
  return {
    ...journal,
    manualIncluded: journal.newManualData !== null,
  };
}

function validateHistoryBytes(
  data: Uint8Array | null,
  label: string,
  performanceTrace: PerformanceTraceSink | undefined,
  phase: string,
): void {
  if (data === null) {
    return;
  }
  const validate = () => {
    try {
      const envelope = decodeSnapshotHistoryEnvelopeWire(new TextDecoder().decode(data));
      validateSnapshotHistoryEnvelope(envelope, { allowUnmigratedPersistedHistory: true });
    } catch (error) {
      throw {
        kind: 'journalCorrupt',
        message: `${label} 无效：${error instanceof Error ? error.message : String(error)}`,
      } satisfies SnapshotImportTransactionError;
    }
  };
  if (performanceTrace === undefined) {
    validate();
  } else {
    performanceTrace.measure('history', phase, validate);
  }
}

function validateManualBytes(data: Uint8Array | null, label: string): void {
  if (data === null) {
    return;
  }
  try {
    decodeManualTrackerEnvelopeWire(new TextDecoder().decode(data));
  } catch (error) {
    throw {
      kind: 'journalCorrupt',
      message: `${label} 无效：${error instanceof Error ? error.message : String(error)}`,
    } satisfies SnapshotImportTransactionError;
  }
}

function encodeOptionalBase64(data: Uint8Array | null): string | null {
  return data === null ? null : bytesToBase64(data);
}

function decodeOptionalBase64(value: string | null): Uint8Array | null {
  return value === null ? null : base64ToBytes(value);
}

function asJournalCorrupt(error: unknown): SnapshotImportTransactionError {
  if (
    typeof error === 'object' &&
    error !== null &&
    'kind' in error &&
    (error as SnapshotImportTransactionError).kind === 'journalCorrupt'
  ) {
    return error as SnapshotImportTransactionError;
  }
  return {
    kind: 'journalCorrupt',
    message: error instanceof Error ? error.message : String(error),
  };
}
